Procedure pointer assignment to dummy argument causes ‘undefined reference’ at link time

The following program does not compile/link with LFortran 0.57

```fortran
module bench_mod
   implicit none
   private
   public :: run

contains

   subroutine run(f)
      implicit none
      interface
         integer function f()
         end function f
      end interface

      procedure(f), pointer :: func_ptr
      integer :: x

      func_ptr => f          ! <-- LFortran fails to resolve this binding
      x = func_ptr()
      print *, "Result:", x
   end subroutine run

end module bench_mod


program main
   use bench_mod
   implicit none

   call run(test_func)

contains

   integer function test_func()
      test_func = 42
   end function test_func

end program main
```

Indeed, this is what I get:
```console
$ lfortran a.f90
Undefined symbols for architecture arm64:
  "_f", referenced from:
      ___module_bench_mod_run in a.tmp_iOXS2Q0zswOqZrPW6K0lWE.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
The command 'clang -o a.out /var/folders/vg/lymvsp1153sgwwfdh3pm45hw0000gn/T/a.tmp_iOXS2Q0zswOqZrPW6K0lWE.o  -L"/Users/ondrej/repos/lfortran/src/bin/../runtime" -Wl,-rpath,"/Users/ondrej/repos/lfortran/src/bin/../runtime" -llfortran_runtime -lm' failed.
Tip: If there is a linker issue, switch the linker using --linker=<CC> option or create an environment variable `export LFORTRAN_LINKER=<CC>`, where CC is clang or gcc
Also, if required use --linker-path=<PATH>, where PATH has location to look for the linker execuatable
```

Thanks for the bug report!

--- a/integration_tests/CMakeLists.txt
+++ b/integration_tests/CMakeLists.txt
@@ -2346,6 +2346,7 @@ RUN(NAME procedure_21 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME procedure_22 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc EXTRAFILES procedure_22_a.f90)
 RUN(NAME procedure_23 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 RUN(NAME procedure_24 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
+RUN(NAME procedure_25 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)
 
 
 RUN(NAME allocated_01 LABELS gfortran llvm llvm_wasm llvm_wasm_emcc)

--- a/integration_tests/procedure_25.f90
+++ b/integration_tests/procedure_25.f90
@@ -1,0 +1,31 @@
+module procedure_25_module
+   implicit none
+contains
+   subroutine run(f)
+      implicit none
+      interface
+         integer function f()
+         end function f
+      end interface
+
+      procedure(f), pointer :: func_ptr
+      integer :: x
+
+      func_ptr => f
+      x = func_ptr()
+      print *, x
+      if(x /= 42) error stop
+   end subroutine run
+end module
+
+
+program procedure_25
+   use procedure_25_module
+   implicit none
+   call run(test_func)
+contains
+   integer function test_func()
+      test_func = 42
+   end function test_func
+
+end program

--- a/src/libasr/codegen/asr_to_llvm.cpp
+++ b/src/libasr/codegen/asr_to_llvm.cpp
@@ -10467,14 +10467,18 @@ public:
             case ASR::symbolType::Variable: {
                 ASR::Variable_t *v = ASR::down_cast<ASR::Variable_t>(x_m_v);
                 fetch_var(v);
-                return ;
+            break;
             }
             case ASR::symbolType::Function: {
-                uint32_t h = get_hash((ASR::asr_t*)x_m_v);
-                if( llvm_symtab_fn.find(h) != llvm_symtab_fn.end() ) {
+                const uint32_t h = get_hash((ASR::asr_t*)x_m_v);
+                if(llvm_symtab_fn_arg.find(h) != llvm_symtab_fn_arg.end()){ // Callback fn arg.
+                    tmp = llvm_symtab_fn_arg[h];
+                } else if( llvm_symtab_fn.find(h) != llvm_symtab_fn.end() ) {
                     tmp = llvm_symtab_fn[h];
+                } else {
+                    throw CodeGenError(std::string("Can't resolve var to Function '") + ASRUtils::symbol_name(x_m_v) + "'");
                 }
-                return;
+            break;
             }
             default: {
                 throw CodeGenError("Only function and variables supported so far");

